core/src/main/java/org/elasticsearch/index/IndexSortConfig.java

```diff
@@ -0,0 +1,218 @@
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.search.MultiValueMode;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Holds all the information that is used to build the sort order of an index.
+ *
+ * The index sort settings are <b>final</b> and can be defined only at index creation.
+ */
+public final class IndexSortConfig {
+    /**
+     * The list of field names
+     */
+    public static final Setting<List<String>> INDEX_SORT_FIELD_SETTING =
+        Setting.listSetting("index.sort.field", Collections.emptyList(),
+            Function.identity(), Setting.Property.IndexScope, Setting.Property.Final);
+
+    /**
+     * The {@link SortOrder} for each specified sort field (ie. <b>asc</b> or <b>desc</b>).
+     */
+    public static final Setting<List<SortOrder>> INDEX_SORT_ORDER_SETTING =
+        Setting.listSetting("index.sort.order", Collections.emptyList(),
+            IndexSortConfig::parseOrderMode, Setting.Property.IndexScope, Setting.Property.Final);
+
+    private static SortOrder parseOrderMode(String value) {
+        try {
+            return SortOrder.fromString(value);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Illegal sort order:" + value);
+        }
+    }
+
+    /**
+     * The {@link MultiValueMode} for each specified sort field (ie. <b>max</b> or <b>min</b>).
+     */
+    public static final Setting<List<MultiValueMode>> INDEX_SORT_MODE_SETTING =
+        Setting.listSetting("index.sort.mode", Collections.emptyList(),
+            IndexSortConfig::parseMultiValueMode, Setting.Property.IndexScope, Setting.Property.Final);
+
+    private static MultiValueMode parseMultiValueMode(String value) {
+        MultiValueMode mode = MultiValueMode.fromString(value);
+        if (mode != MultiValueMode.MAX && mode != MultiValueMode.MIN) {
+            throw new IllegalArgumentException("Illegal index sort mode:[" + mode + "], " +
+                "must be one of [" + MultiValueMode.MAX + ", " + MultiValueMode.MIN + "]");
+        }
+        return mode;
+    }
+
+    /**
+     * The missing value for each specified sort field (ie. <b>_first</b> or <b>_last</b>)
+     */
+    public static final Setting<List<String>> INDEX_SORT_MISSING_SETTING =
+        Setting.listSetting("index.sort.missing", Collections.emptyList(),
+            IndexSortConfig::validateMissingValue, Setting.Property.IndexScope, Setting.Property.Final);
+
+    private static String validateMissingValue(String missing) {
+        if ("_last".equals(missing) == false && "_first".equals(missing) == false) {
+            throw new IllegalArgumentException("Illegal missing value:[" + missing + "], " +
+                "must be one of [_last, _first]");
+        }
+        return missing;
+    }
+
+    final String[] fields;
```

Comment:
I think it is also worth leaving a comment about how this is stored like this for easy reading from the settings. It looks funny to my java-accustomed eye.